### PR TITLE
Modernize Administrator Queries page

### PR DESCRIPTION
**Pull request by mylesw27**

### CODE REVIEW

File: apps/prairielearn/src/pages/administratorQueries/administratorQueries.html.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+            <div class="card-header bg-primary text-white">Queries</div>
+            <div class="table-responsive">
+              <table class="table table-sm table-hover table-striped">
+                <tbody>
+                  ${queries.map(
+                    (query) => html`
+                      <tr>
+                        <td>
+                          <a href=${`../administrator/query/${query.link}`}>
+                            <code>${query.sqlFilename}</code>
```
Comment (mylesw27) [inline apps/prairielearn/src/pages/administratorQueries/administratorQueries.html.ts:41]: The <tt> tag is deprecated (per MDN docs, https://developer.mozilla.org/en-US/docs/Web/HTML/Element/tt). Replaced with a <code> tag.

File: apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts [hunk 2]
```diff
@@ -14,16 +15,21 @@ router.get(
   asyncHandler(async (req, res, _next) => {
     const fileList = await fsPromises.readdir(queriesDir);
     const jsonList = _.filter(fileList, (f) => /\.json$/.test(f));
-    res.locals.queries = await Promise.all(
+    const queries = await Promise.all(
       jsonList.map(async (f) => {
         const contents = await jsonLoad.readJSONAsync(path.join(queriesDir, f));
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts:20]: I'd be inclined to define a Zod schema for the loaded JSON so that we're always validating that is has the expected properties and shape.

File: apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts [hunk 3]
```diff
@@ -14,16 +15,21 @@ router.get(
   asyncHandler(async (req, res, _next) => {
     const fileList = await fsPromises.readdir(queriesDir);
     const jsonList = _.filter(fileList, (f) => /\.json$/.test(f));
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts:17]: I like to avoid [Lodash](https://lodash.com/docs/4.17.15) in cases where there's a native JS equivalent. In this case, you can use [`Array.prototype.filter`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) directly.

File: apps/prairielearn/src/pages/administratorQueries/administratorQueries.html.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+          <div class="card mb-4">
+            <div class="card-header bg-primary text-white">Queries</div>
+            <div class="table-responsive">
+              <table class="table table-sm table-hover table-striped">
+                <tbody>
+                  ${queries.map(
+                    (query) => html`
+                      <tr>
+                        <td>
+                          <a href=${`../administrator/query/${query.link}`}>
```
[the remark this answers is not in the document]
Comment (mylesw27) [reply apps/prairielearn/src/pages/administratorQueries/administratorQueries.html.ts]: Resolved in commit a1e7740.

File: apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts (hunk 2 above)
Comment (mylesw27) [reply apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts]: Resolved in commit a1e7740.

File: apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts (hunk 3 above)
Comment (mylesw27) [reply apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts]: Resolved in commit a1e7740.

File: apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts [hunk 5]
```diff
@@ -0,0 +1,60 @@
+import { z } from 'zod';
+import express = require('express');
+const router = express.Router();
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts:3]: We usually relocate `const router = ...` out of the middle of the `import` statements when refactoring. That is, move this below all the other `import` statements.

File: apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+import asyncHandler = require('express-async-handler');
+import fsPromises = require('node:fs/promises');
+import path = require('path');
```
Comment (jonatanschroeder) [inline apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts:6]: I believe all these imports allow `import from` instead of `import = require`.

Comment (mylesw27) [reply apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts:6]: screenshot_2023-10-18_at_2 39 36_pm
Unfortunately, we are unable to use `import from` instead of `import = require`.

File: apps/prairielearn/src/lib/db-types.ts
[the remark this answers is not in the document]
Comment (mylesw27) [reply apps/prairielearn/src/lib/db-types.ts:1]: Resolved in commit 4d28c98.

File: apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+  resultFormats: z.any().nullable(),
+  comment: z.any().nullable(),
+  params: z.array(AdministratorQueryParamsSchema).nullable().optional(),
+});
+export type AdministratorQueryJson = z.infer<typeof AdministratorQueryJsonSchema>;
+
+export type AdministratorQuery = AdministratorQueryJson & {
+  sqlFilename: string;
+  link: string;
+};
```
[the remark this answers is not in the document]
Comment (mylesw27) [reply apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts]: Resolved in commit 4d28c98.

File: apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+  '/',
+  asyncHandler(async (req, res, _next) => {
+    const fileList = await fsPromises.readdir(queriesDir);
+    const jsonList = fileList.filter((f) => /\.json$/.test(f));
+    const queries = await Promise.all(
+      jsonList.map(async (f) => {
+        const contents = AdministratorQueryJsonSchema.parse(
+          await jsonLoad.readJSONAsync(path.join(queriesDir, f)),
+        );
+        console.log('new query', f, contents);
```
[the remark this answers is not in the document]
Comment (mylesw27) [reply apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts]: Resolved in commit 4d28c98.

File: apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+  default: z.string().nullable().optional(),
+  comment: z.any().nullable(),
+});
+
+const AdministratorQueryJsonSchema = z.object({
+  description: z.string(),
+  resultFormats: z.any().nullable(),
+  comment: z.any().nullable(),
+  params: z.array(AdministratorQueryParamsSchema).nullable().optional(),
+});
```
[the remark this answers is not in the document]
Comment (mylesw27) [reply apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts]: Resolved in commit 4d28c98.

File: apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts (hunk 6 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts]: Try `import * as express from 'express'`

Comment (mylesw27) [reply apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts]: That worked. Thanks for the help, @jonatanschroeder !

Resolved in commit 3bbab10.

File: apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+          ...contents,
+          sqlFilename: f.replace(/\.json$/, '.sql'),
+          link: f.replace(/\.json$/, ''),
+        };
+      }),
+    );
+    res.send(
+      AdministratorQueries({
+        resLocals: res.locals,
+        queries: queries as AdministratorQuery[],
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts:47]: Having to do `as ...` is a bit of a code smell - why was this necessary?

File: apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts (hunk 6 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts]: Happy to help. BTW, I believe you missed one more require further down.

File: apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts (hunk 10 above)
Comment (mylesw27) [reply apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts]: Code smell was the result of using `any` types. Added a zod schema in commit 782f18f to address the shape of the queries and remove the use of `any` types.

File: apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts (hunk 6 above)
Comment (mylesw27) [reply apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts]: You are correct. This is now resolved in commit 782f18f.

Also, I had issues with importing the asyncHandler on line 3 and reverted this back to `asyncHandler = require(...`.

File: apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts (hunk 10 above)
Comment (nwalters512) [reply apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts]: I believe this change now makes the page throw an error on load. I'd recommend always doing a quick manual functionality test before pushing new commits.

File: apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+          );
+          return {
+            ...contents,
+            sqlFilename: f.replace(/\.json$/, '.sql'),
+            link: f.replace(/\.json$/, ''),
+          };
+        }),
+      ),
+    );
+    console.log(queries);
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts:71]:
```suggestion
```

File: apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts (hunk 10 above)
Comment (mylesw27) [reply apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts]: Resolved in commit 2a7e7b8.

File: apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts (hunk 11 above)
Comment (mylesw27) [reply apps/prairielearn/src/pages/administratorQueries/administratorQueries.ts]: Removed in commit 2a7e7b8.

**Review by nwalters512 — APPROVED** [review]
